The big refactor/fix branch

miceg opened the pull request:
Discussed in #8; this has a bunch of refactoring and fixes that are difficult to untangle.

AppState changes:

* Construct an "OK" response with `x-vercel-verify` header on start-up, and store it in the `AppState`, so it can be cloned as needed.

  This avoids needing an `unwrap()` in the ingest handler, that could be triggered with a configuration error but only after the first request.

* Accept `vercel_secret` as a `&[u8]`, and move all the HMAC handling in there.

Ingest handler:

* Respond to ingest requests without a signature with HTTP 200 OK and `x-vercel-verify` header, to match Vercel not sending this header, even when you set a custom secret during setup.

  Added a test case for this and a request with a valid signature but empty payload (in case they fix this).

* Verify the request payload's HMAC before trying to decode it as UTF-8.

* Return Unauthorized error whenever `x-vercel-signature` header is present, and is not in the expected format or doesn't pass HMAC checks, and don't send a `x-vercel-verify` header in those responses.

* Avoid copying the payload by using `str::from_utf8()` on a `&[u8]`, rather than `::to_vec()` and passing it into a `String`.

* Use `IntoResponse::into_response()` to construct responses with empty bodies and no special header.

* Return Unprocessable Entity error whenever it doesn't parse correctly, rather than returning OK.

* Removed many `unwrap()` calls which could crash the binary with a crafted request.

* Added test cases for the expected response codes, and also for some requests that could crash because of `unwrap()`.

Other handlers:

* Make `root()` return a `StatusCode` directly, like `health_check()`.

* Fix needless-returns.

Tests:

* Use `?` operator rather than `unwrap()` where possible.

* Use arrays for constants rather than `vec!`.

* Use `&str` and `b""` for constants where possible.

miceg committed:
Big refactor / cleanups:

AppState changes:

* Construct an "OK" response with `x-vercel-verify` header on start-up, and
  store it in the `AppState`, so it can be cloned as needed.

  This avoids needing an `unwrap()` in the ingest handler, that could be
  triggered with a configuration error but only after the first request.

Ingest handler fixes:

* Respond to ingest requests without a signature with HTTP 200 OK and
  `x-vercel-verify` header, to match Vercel not sending this header, even
  when you set a custom secret during setup.

  Added a test case for this and a request with a valid signature but empty
  payload.

* Verify the request payload before trying to decode it as UTF-8.

* Return Unauthorized error whenever `x-vercel-signature` header is present,
  and is not in the expected format or doesn't pass HMAC checks, and don't
  send a `x-vercel-verify` header in those responses.

* Avoid copying the payload by using `str::from_utf8()` on a `&[u8]`, rather
  than `::to_vec()` and passing it into a `String`.

* Use `IntoResponse::into_response()` to construct responses with empty
  bodies and no special header.

* Return Unprocessable Entity error whenever it doesn't parse correctly,
  rather than returning OK.

* Removed many `unwrap()` calls which could crash the binary with a crafted
  request.

* Added test cases for the expected response codes, and also for some
  requests that could crash because of `unwrap()`.

Other handlers:

* Make `root()` return a `StatusCode` directly, like `health_check()`.

* Fix needless-returns.

Tests:

* Use `?` operator rather than `unwrap()` where possible.

* Use arrays for constants rather than `vec!`.

* Use `&str` and `b""` for constants where possible.

miceg committed:
Fix clippy lint

dacbd commented on src/main.rs:
```diff
@@ -94,14 +94,14 @@ async fn main() -> anyhow::Result<()> {
-        vercel_secret: hmac::Key::new(
+    let state = types::AppState::new(
+        &args.vercel_verify,
+        hmac::Key::new(
             hmac::HMAC_SHA1_FOR_LEGACY_USE_ONLY,
             args.vercel_secret.as_bytes(),
         ),
-        log_queue: tx,
-    };
+        tx,
+    )?;
```
lets construct the `hmac::Key` in the new function, so you can just pass it another string

miceg replied:
Done

dacbd commented on src/handlers.rs:
```diff
@@ -27,45 +24,51 @@ pub async fn ingest(
-                .expect("Defined Responses to be infalliable.");
-        }
+    let Some(sig_header) = headers.get("x-vercel-signature") else {
+        // Vercel's verification request isn't signed, even if you set a custom
+        // secret in advance. It also expect 200 OK.
+        //
+        // Configure Vercel to send a random custom static header and
+        // configure your HTTPS load balancer in front of vercel-log-drain
+        // to require it, if you want to discard bot traffic.
```
```suggestion
```

let's just remove this comment since it is documented in the recommended setup process.

miceg replied:
Done

dacbd commented on src/handlers.rs:
```diff
@@ -79,11 +82,9 @@ pub async fn ingest(
-        .body(Body::empty())
-        .expect("Defined Responses to be infalliable.");
+
+    state.ok_response()
```
```suggestion
    return state.ok_response();
```
I'd prefer an explicit return in this case.

miceg replied:
Done

miceg committed:
Addressing review feedback:

* Remove comment about signature handling, because that's in the README.

* Add explicit `return` where a function also has a `return` elsewhere.

* Move `hmac::Key` construction into `AppState::new`, which now accepts a
  key as `&[u8]`.

* Add `AppState::sign_request_for_test_only` for signing requests, which is
  only for test cases.

This means that all the HMAC code is in one place.

miceg committed:
fmt

dacbd reviewed (approved):
This is great, Thanks a bunch! If you have some spare time I'd love it if you reached out, I'd like to ask you a few rust questions sometime. (dabarnes2b@gmail.com)
